File: dp/memoization/gridTraveler.py
def grid_traveler(n, m, memo={}):
    key = f"{n},{m}"  # if n > m else f"{m},{n}"
    if key in memo:
        return memo[key]
    if n == 1 and m == 1:
        return 1
    if n == 0 or m == 0:
        return 0
    memo[key] = grid_traveler(n - 1, m) + grid_traveler(n, m - 1)
    return memo[key]


print(grid_traveler(300, 300))

File: dp/tabulation/gridTraveler.js
const makeGrid = (n, m) => {
  const grid = [];
  for (let i = 0; i <= n; i++) {
    grid.push(Array(m + 1).fill(0));
  }
  return grid;
};

const gridTraveler = (n, m) => {
  const grid = makeGrid(n, m);
  grid[1][1] = 1;
  for (let i = 0; i <= n; i++) {
    for (let j = 0; j <= m; j++) {
      if (i <= n - 1) grid[i + 1][j] += grid[i][j];
      if (j <= n - 1) grid[i][j + 1] += grid[i][j];
    }
  }
  return grid[n][m];
};
console.log(gridTraveler(30, 30));
